src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalConnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_DEBUG("Already connected - " << this->DeviceName);
+    return PLUS_SUCCESS;
+  }
+
+  if (this->InputChannels.empty())
+  {
+    LOG_WARNING("No force input has been provided");
+  }
+  else
+  {
+    if (this->InputChannels.size() > 1)
+    {
+      LOG_WARNING("Multiple input channels present.  Only first channel will be checked for force data");
+    }
+    vtkPlusDataSource* forceInput;
+    if (!this->InputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))
+    {
+      LOG_WARNING("No Force tool in input channel.  Forces will not be available");
+    }
+  }
+
+  HDErrorInfo errorFlush;
+  while (HD_DEVICE_ERROR(errorFlush = hdGetError())) {}
+
+  DeviceHandle = hdInitDevice(this->DeviceName.c_str());
+
+  HDErrorInfo error;
+  if (HD_DEVICE_ERROR(error = hdGetError()))
+  {
+    LOG_ERROR( "Failed to initialize Phantom Omni " << this->DeviceName);
+    DeviceHandle = -1;
+    return PLUS_FAIL;
+  }
+
+  // Enable forces
+  hdEnable(HD_FORCE_OUTPUT);
+  hdEnable(HD_FORCE_RAMPING);
+
+  this->isHapticDeviceInitialized = true;
+  LOG_INFO("Phantom initialized: " << this->DeviceName)
+
+  hdStartScheduler();
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalDisconnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    hdDisableDevice(DeviceHandle);
+    hdStopScheduler();
+    this->isHapticDeviceInitialized=false;
+  }
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::ReadConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, rootConfigElement);
+  XML_READ_CSTRING_ATTRIBUTE_REQUIRED(DeviceName,deviceConfig);
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(trackerConfig, rootConfigElement);
+  trackerConfig->SetAttribute("DeviceName", this->DeviceName.c_str());
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStartRecording()
+{
+  if (!this->isHapticDeviceInitialized)
+  {
+    LOG_ERROR("InternalStartRecording failed: OpenHaptics device has not been initialized");
+    return PLUS_FAIL;
+  }
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStartRecording");
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStopRecording()
+{
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStopRecording");
+  return PLUS_SUCCESS;
+}
+
+HDCallbackCode HDCALLBACK
+vtkPlusOpenHapticsDevice::positionCallback(void* pData)
+{
+  
+  vtkPlusOpenHapticsDevice* client = reinterpret_cast<vtkPlusOpenHapticsDevice*>(pData);
+  HHD handle = client->DeviceHandle;
+  HD_state state = client->DeviceState;
+
+  const double unfilteredTimestamp = vtkPlusAccurateTimer::GetSystemTime();
+  ++client->FrameNumber;

```

Comment:
Not really a big deal, but this will make it so that frame 0 never exists. Again, this is not wrong, just fyi.